keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1587 @@
+  not exist):
+  - When a CRD is created/updated: The type check errors are detected by an
+    control loop watching the CRDs with an informer in the API server, and
+    reported in the status of the ValidatingAdmissionPolicy.
+  - When a policy is validated: runtime evaluations of the expression are
+    skipped, and the validation is reported as failed, with a type check error.
+
+The main disadvantage of this approach is that CEL rules that would have
+evaluated successfully dynamtically (where type checking is skipped), will
+instead fail. By favoring type safety here were are trading off avaiability for
+consistent and predictable behavior.
+
+Example: Typesafe access to object 
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "validate-xyz.example.com"
+spec:
+  constraints:
+   rules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["deployments"]
+      scope:       "*"
+  validations:
+    # replicas is accessible because this resource matches only v1 deployments
+  - rule: "object.spec.replicas < 100"
+```
+
+Example: Typesafe access only to metadata
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "validate-xyz.example.com"
+spec:
+  constraints:
+   rules:
+    - apiGroups:   ["*"]
+      apiVersions: ["*"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["*"]
+      scope:       "*"
+  validations:
+    # minReadySeconds is not accessible because this resource matches multiple types
+  - rule: "object.spec.minReadySeconds > 60" # ERROR! Not such field "minReadySeconds".
+    # metadata is always accessible
+  - rule: "object.name.startsWith('xyz')"
+```
+
+Alternative Considered: Do not use CEL type-checking and instead dynamically evaluate
+all CEL expressions.
+
+Pros:
+
+- Can write a single CEL expression that processes multiple GVKs
+
+Cons:
+
+- All errors are reported during admission control validation. If instead
+  type-checking is used, CEL type-check errors can be surfaced much earlier in
+  the status of `ValidatingAdmissionPolicy` resources when the
+  `ValidatingAdmissionPolicy`, configuration CRDs and configuration resources
+  are written to the API Server. This provides a workflow where type-checking
+  can be used to detect errors early, preventing them from occurring during
+  admission control validation and making it safer to operate a policy using
+  `failureMode: Fail`.
+
+#### Failure Policy
+
+For in-process validation there is no remote request, so errors should be
+deterministic. We believe this significantly reduces the risk of "fail closed"
+admission control as compared to webhooks.
+
+Because failure policy is most often selected based on the need to guarantee
+enforcement, we propose defaulting failure policy to "fail" and allowing it to
```

Comment:
Would it make sense to default to fail open until we have a way to ensure a policy isn't going to break the cluster? e.g. https://github.com/kubernetes/kubernetes/issues/9215

Reply:
I've gone back and forth on this one.  One one hand, defaulting to fail open can make rollouts safer, one the other hand, it's a problem for compliance where defaulting to fail closed is very much needed.  Webhooks already default to fail closed, so I'm tempted to take that path purely as an API consistency choice-- we are claiming that fail closed is less dangerous with CEL.. Curious what others think.

Reply:
This is covered in the match criteria section.  We intend to support exclusion on each major match field. How we do this varies by field. For example scope has only two options (cluster, namespaced) so we don't need a exclude for that field, just  use `scope: [Namespaced]`, but for unbounded fields there needs to be an explicit exclude capability.